Source/Core/Core/HW/GBACore.cpp

```diff
@@ -0,0 +1,518 @@
+        continue;
+
+      vf = VFileMemChunk(buffer.data(), info.uncompressed_size);
+      if (mCoreIsCompatible(vf) == mPLATFORM_GBA)
+      {
+        vf->seek(vf, 0, SEEK_SET);
+        break;
+      }
+
+      vf->close(vf);
+      vf = nullptr;
+    } while (unzGoToNextFile(zip) == UNZ_OK);
+    unzClose(zip);
+  }
+  return vf;
+}
+
+bool Core::LoadROM(const char* rom_path)
+{
+  VFile* vf{};
+
+  vf = LoadROM_Archive(rom_path);
+  if (!vf)
+    vf = LoadROM_Zip(rom_path);
+  if (!vf)
+  {
+    vf = VFileOpen(rom_path, O_RDONLY);
+    if (vf)
+    {
+      if (mCoreIsCompatible(vf) != mPLATFORM_GBA)
+      {
+        vf->close(vf);
+        vf = nullptr;
+      }
+      else
+      {
+        vf->seek(vf, 0, SEEK_SET);
+      }
+    }
+  }
+
+  if (!vf)
+  {
+    PanicAlertFmtT("Error: GBA{0} failed to open the ROM in {1}", m_device_number + 1, rom_path);
+    return false;
+  }
+
+  if (!m_core->loadROM(m_core, vf))
+  {
+    PanicAlertFmtT("Error: GBA{0} failed to load the ROM in {1}", m_device_number + 1, rom_path);
+    vf->close(vf);
+    return false;
+  }
+
+  return true;
+}
+
+bool Core::LoadSave(const char* save_path)
+{
+  VFile* vf = VFileOpen(save_path, O_CREAT | O_RDWR);
+  if (!vf)
+  {
+    PanicAlertFmtT("Error: GBA{0} failed to open the save in {1}", m_device_number + 1, save_path);
+    return false;
+  }
+
+  if (!m_core->loadSave(m_core, vf))
+  {
+    PanicAlertFmtT("Error: GBA{0} failed to load the save in {1}", m_device_number + 1, save_path);
+    vf->close(vf);
+    return false;
+  }
+
+  return true;
+}
+
+void Core::SetSIODriver()
+{
+  GBASIOJOYCreate(&m_sio_driver);
+  GBASIOSetDriver(&static_cast<::GBA*>(m_core->board)->sio, &m_sio_driver, SIO_JOYBUS);
+
+  m_sio_driver.core = this;
+  m_sio_driver.load = [](GBASIODriver* driver) {
+    static_cast<SIODriver*>(driver)->core->m_link_enabled = true;
+    return true;
+  };
+  m_sio_driver.unload = [](GBASIODriver* driver) {
+    static_cast<SIODriver*>(driver)->core->m_link_enabled = false;
+    return true;
+  };
+}
+
+void Core::SetCallbacks()
+{
+  mCoreCallbacks callbacks{};
+  callbacks.context = this;
+  callbacks.keysRead = [](void* context) { reinterpret_cast<Core*>(context)->OnKeysRead(); };
+  callbacks.videoFrameEnded = [](void* context) {
+    auto core = reinterpret_cast<Core*>(context);
+    core->m_frontend->FrameEnded(core->m_video_buffer);
+  };
+  m_core->addCoreCallbacks(m_core, &callbacks);
+}
+
+void Core::SetAVStream()
+{
+  m_stream.core = this;
+  m_stream.videoDimensionsChanged = nullptr;
+  m_stream.postVideoFrame = nullptr;
+  m_stream.postAudioFrame = nullptr;
+  m_stream.postAudioBuffer = [](struct mAVStream* stream, struct blip_t* left,
+                                struct blip_t* right) {
+    auto device = static_cast<AVStream*>(stream)->core;
+    std::vector<s16> buffer(SAMPLES * 2);
+    blip_read_samples(left, &buffer[0], SAMPLES, 1);
+    blip_read_samples(right, &buffer[1], SAMPLES, 1);
```

Comment:
This looks weird to me, isn't this reading into `buffer[0]` with `left`, overwrite `buffer[1..512]` with `right` and leave `buffer[513..]` unused?

Reply:
No, the last parameter tells it to only write every other location (it's the `stereo` parameter that tells it the buffer is interlaced)

Reply:
Oh, got it. Thanks 👍